File: src/cli/root-query/parse-root-queries.ts
import { parseRootQuery } from "./parse-root-query.js";
import type { ParsedRootQuery } from "./parsed-root-query.js";

export type RootQueryParseSuccess = Readonly<{
  ok: true;
  queries: readonly ParsedRootQuery[];
}>;

export type RootQueryParseFailure = Readonly<{
  ok: false;
  error: string;
}>;

export type RootQueryParseResult =
  | RootQueryParseSuccess
  | RootQueryParseFailure;

export function parseRootQueries(value: string): RootQueryParseResult {
  if (value === "") {
    return { ok: false, error: "empty --roots value" };
  }
  const tokens = value.split(",");
  const queries: ParsedRootQuery[] = [];
  for (const token of tokens) {
    const r = parseRootQuery(token);
    if ("error" in r) {
      return { ok: false, error: r.error };
    }
    queries.push(r);
  }
  return { ok: true, queries };
}

File: crates/unsnarl-root-query/src/parse_root_queries.rs
use crate::parse_root_query::parse_root_query;
use crate::parsed_root_query::ParsedRootQuery;

pub fn parse_root_queries(value: &str) -> Result<Vec<ParsedRootQuery>, String> {
    if value.is_empty() {
        return Err("empty --roots value".to_string());
    }
    let mut queries: Vec<ParsedRootQuery> = Vec::new();
    for token in value.split(',') {
        queries.push(parse_root_query(token)?);
    }
    Ok(queries)
}

#[cfg(test)]
#[path = "parse_root_queries_test.rs"]
mod parse_root_queries_test;
